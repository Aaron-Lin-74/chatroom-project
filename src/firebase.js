import { initializeApp } from 'firebase/app'
import { useState, useEffect } from 'react'
import {
  getFirestore,
  query,
  collection,
  orderBy,
  limit,
  addDoc,
  updateDoc,
  Timestamp,
  serverTimestamp,
  onSnapshot,
} from 'firebase/firestore'

import {
  getAuth,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithPopup,
  signInWithRedirect,
  onAuthStateChanged,
  GoogleAuthProvider,
  signOut,
  updateProfile,
} from 'firebase/auth'

import {
  getStorage,
  ref,
  uploadBytesResumable,
  getDownloadURL,
} from 'firebase/storage'

const firebaseConfig = {
  apiKey: 'AIzaSyDJBHpBTyk0rq2_j3-9Yywg-x9_wvUQNnQ',
  authDomain: 'auth-development-3eefa.firebaseapp.com',
  projectId: 'auth-development-3eefa',
  storageBucket: 'auth-development-3eefa.appspot.com',
  messagingSenderId: '553742939413',
  appId: '1:553742939413:web:9a658201c3ebe29565251c',
  measurementId: '${config.measurementId}',
}
const app = initializeApp(firebaseConfig)
const db = getFirestore(app)
const auth = getAuth()
const provider = new GoogleAuthProvider()
const messagesRef = collection(db, 'messages')

const getMessageQuery = () => {
  const q = query(messagesRef, orderBy('createdAt'), limit(50))
  return q
}

const signInWithGoogle = async () => {
  // signInWithRedirect(auth, provider)
  try {
    await signInWithPopup(auth, provider)
  } catch (err) {
    console.log(err)
  }
}

const signOutUser = () => {
  // Sign out of Firebase.
  signOut(auth)
}

// Returns true if a user is signed-in.
const isUserSignedIn = () => {
  return !!auth.currentUser
}

// Returns the signed-in user's profile Pic URL.
function getProfilePicUrl() {
  return auth.currentUser.photoURL || '/images/default_profile.jpg'
}

// Returns the signed-in user's display name.
function getUserName() {
  return auth.currentUser.displayName
}

export {
  auth,
  db,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithGoogle,
  getProfilePicUrl,
  isUserSignedIn,
  getUserName,
  signOutUser,
  messagesRef,
  getMessageQuery,
  onSnapshot,
  updateProfile,
}

// custome hook to get the current user
export function useAuth() {
  const [currentUser, setCurrentUser] = useState(null)
  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => setCurrentUser(user))
    return unsubscribe
  }, [])
  return currentUser
}
