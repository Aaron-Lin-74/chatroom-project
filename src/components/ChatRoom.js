import React, { useRef } from 'react'
import { getMessageQuery } from '../firebase'
import MessagePanel from './MessagePanel'
import ChatMessage from './ChatMessage'
import { useCollectionData } from 'react-firebase-hooks/firestore'

function ChatRoom() {
  const dummy = useRef()

  const q = getMessageQuery()
  const [messages] = useCollectionData(q, { idField: 'id' })

  return (
    <>
      <main>
        {messages &&
          messages.map((msg) => <ChatMessage key={msg.id} message={msg} />)}

        <span ref={dummy}></span>
      </main>

      <MessagePanel />
    </>
  )
}
export default ChatRoom
