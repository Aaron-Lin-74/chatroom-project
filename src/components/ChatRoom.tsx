import React, { useState, useEffect, useRef } from 'react';
import { getMessageQuery, onSnapshot } from '../firebase';
import MessagePanel from './MessagePanel';
import ChatMessage from './ChatMessage';
import { MessageType } from '../types/message';

function ChatRoom() {
  const [messages, setMessages] = useState<MessageType[]>([]);
  const [messageNumber, setMessageNumber] = useState(25);
  const spanRef = useRef<HTMLSpanElement>(null);
  const chatBoxRef = useRef<HTMLDivElement>(null);

  // Check the user has scroll to the top of the chat box
  const checkScrollTop = (e: Event) => {
    const targetDiv = e.target as HTMLDivElement;
    if (targetDiv.scrollTop === 0) {
      // Display the show more clickable
      if (spanRef.current !== null) {
        spanRef.current.style.display = 'block';
      }
    }
  };

  // When the user click show more, load 25 more old messages
  const showMoreMessages = () => {
    setMessageNumber((messageNumber) => (messageNumber += 25));
    if (spanRef.current !== null) {
      spanRef.current.style.display = 'none';
    }
  };

  useEffect(() => {
    // Check whether to load the extra messages when scroll to top
    const chatBox = chatBoxRef.current;
    if (chatBox !== null) {
      chatBox.addEventListener('scroll', checkScrollTop);
    }

    return () => {
      if (chatBox !== null) {
        chatBox.removeEventListener('scroll', checkScrollTop);
      }
    };
  }, []);

  // Start listening to the query.
  useEffect(() => {
    // Create the query to load the last n messages and listen for new ones.
    const query = getMessageQuery(messageNumber);

    // If in the same day, just show the hour:min, otherwise show the date and time
    const displayTime = (date: Date) => {
      const now = new Date();
      if (isSameDay(date, now)) {
        return formatTime(date.getHours(), date.getMinutes());
      }
      return `${date.toDateString()} ${formatTime(
        date.getHours(),
        date.getMinutes()
      )}`;
    };
    const unsubscribe = onSnapshot(query, function (snapshot) {
      const descMessages: MessageType[] = snapshot.docs.map((doc) => {
        // Add createdAt, and unique id properties to the message data object
        return {
          ...doc.data(),
          photoURL: doc.data().photoURL,
          uid: doc.data().uid,
          createdAt: displayTime(doc.data().createdAt.toDate()),
          id: doc.id,
        };
      });
      // since the lastest n messages are in descend order, so we need to reverse
      setMessages(descMessages.reverse());
      scrollDown();
    });
    return unsubscribe;
  }, [messageNumber]);

  // Scroll to the bottom of the chatbox
  const scrollDown = () => {
    if (chatBoxRef.current !== null) {
      chatBoxRef.current.scrollTop = chatBoxRef.current.scrollHeight;
    }
  };

  // Change the time format to '00:00'
  const formatTime = (hour: number, minute: number): string => {
    return `${hour < 10 ? '0' + hour : hour}:${
      minute < 10 ? '0' + minute : minute
    }`;
  };

  // return true is two dates are of the same day
  const isSameDay = (date1: Date, date2: Date): boolean => {
    if (
      date1.getFullYear() === date2.getFullYear() &&
      date1.getMonth() === date2.getMonth() &&
      date1.getDate() === date2.getDate()
    ) {
      return true;
    }
    return false;
  };
  return (
    <main id='chatBox' ref={chatBoxRef}>
      <span id='show-more-message' ref={spanRef} onClick={showMoreMessages}>
        _______ Show more ________
      </span>
      <div className='message-box'>
        {messages &&
          messages.map((msg) => {
            return (
              <ChatMessage key={msg.id} message={msg} scrollDown={scrollDown} />
            );
          })}
      </div>
      <MessagePanel />
    </main>
  );
}
export default ChatRoom;
