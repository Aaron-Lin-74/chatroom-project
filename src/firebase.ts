import { initializeApp } from 'firebase/app';
import { useState, useEffect } from 'react';
import {
  getFirestore,
  query,
  collection,
  orderBy,
  limit,
  getDocs,
  where,
  updateDoc,
  onSnapshot,
} from 'firebase/firestore';

import {
  User as FirebaseUser,
  getAuth,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithPopup,
  onAuthStateChanged,
  GoogleAuthProvider,
  signOut,
  updateProfile,
} from 'firebase/auth';

const firebaseConfig = {
  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.REACT_APP_FIREBASE_APP_ID,
  measurementId: 'G-N1MGDFF41J',
};
const app = initializeApp(firebaseConfig);
const db = getFirestore(app);
const auth = getAuth();
const provider = new GoogleAuthProvider();
const messagesRef = collection(db, 'messages');

const getMessageQuery = (n: number) => {
  const q = query(messagesRef, orderBy('createdAt', 'desc'), limit(n));
  return q;
};

const signInWithGoogle = async () => {
  // signInWithRedirect(auth, provider)
  try {
    await signInWithPopup(auth, provider);
  } catch (err) {
    console.log(err);
  }
};

const signOutUser = () => {
  // Sign out of Firebase.
  signOut(auth);
};

// Returns true if a user is signed-in.
const isUserSignedIn = () => {
  return !!auth.currentUser;
};

// Returns the signed-in user's profile Pic URL.
function getProfilePicUrl() {
  return auth.currentUser?.photoURL || '/images/default_profile.jpg';
}

// Returns the signed-in user's display name.
function getUserName() {
  return auth.currentUser?.displayName || 'User';
}

/** Update all messages' profile image that were sent by this signed-in user.
 * @param uid - User id
 * @param url - The new profile image url
 */
async function getMessagesAndUpdatePhoto(uid: string, url: string) {
  try {
    const q = query(messagesRef, where('uid', '==', uid));
    const querySnapshot = await getDocs(q);
    // could not await in side the forEach loop, use map instead
    // querySnapshot.forEach((doc) => {
    // doc.data() is never undefined for query doc snapshots
    // updateDoc(doc.ref, {
    //   photoURL,
    // })
    // })
    await Promise.all(
      querySnapshot.docs.map(async (doc) => {
        await updateDoc(doc.ref, { photoURL: url });
      })
    );
  } catch (err) {
    console.error(err);
  }
}

export {
  auth,
  db,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithGoogle,
  getProfilePicUrl,
  isUserSignedIn,
  getUserName,
  signOutUser,
  messagesRef,
  getMessageQuery,
  onSnapshot,
  updateProfile,
  getMessagesAndUpdatePhoto,
};

/** custome hook to get the current user. */
export function useAuth() {
  const [currentUser, setCurrentUser] = useState<FirebaseUser | null>(null);
  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) =>
      setCurrentUser(user)
    );
    return unsubscribe;
  }, []);
  return currentUser;
}
