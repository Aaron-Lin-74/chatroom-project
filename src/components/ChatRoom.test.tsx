import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ChatRoom from './ChatRoom';
import { enableNetwork, disableNetwork } from 'firebase/firestore';
import { db } from '../firebase';

describe('ChatRoom component', () => {
  beforeEach(async () => {
    await enableNetwork(db);
  });

  afterEach(async () => {
    await disableNetwork(db);
  });

  test('should match the snapshot', () => {
    const { asFragment } = render(<ChatRoom />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render the component', () => {
    render(<ChatRoom />);
    screen.getByText(/show more/i);
  });

  test('should hide the span of show more when click it', () => {
    render(<ChatRoom />);
    fireEvent.scroll(screen.getByRole('main'), { target: { scrollY: 0 } });
    const spanElement = screen.getByText(/show more/i);
    fireEvent.click(spanElement);
    expect(spanElement).toHaveStyle('display: none');
  });

  test('should show the span of show more when scroll top', () => {
    render(<ChatRoom />);
    const spanElement = screen.getByText(/show more/i);
    fireEvent.click(spanElement);
    expect(spanElement).toHaveStyle('display: none');
    fireEvent.scroll(screen.getByRole('main'), {
      target: { scrollY: 0 },
    });
    expect(spanElement).toHaveStyle('display: block');
  });
});
