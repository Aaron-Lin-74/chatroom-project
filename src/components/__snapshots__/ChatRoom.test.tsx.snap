// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoom component should match the snapshot 1`] = `
<DocumentFragment>
  <main
    id="chatBox"
  >
    <span
      id="show-more-message"
    >
      _______ Show more ________
    </span>
    <div
      class="message-box"
    />
    <div
      class="message-panel"
    >
      <form
        class="message-panel-form"
      >
        <input
          aria-label="Type message here"
          class="msg-input"
          placeholder="Type your message and hit enter"
          value=""
        />
        <button
          aria-label="Add an emoji"
          class="chat-btns"
          id="emoji-btn"
          title="Add an emoji"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
            />
            <path
              d="M4.285 9.567a.5.5 0 0 1 .683.183A3.498 3.498 0 0 0 8 11.5a3.498 3.498 0 0 0 3.032-1.75.5.5 0 1 1 .866.5A4.498 4.498 0 0 1 8 12.5a4.498 4.498 0 0 1-3.898-2.25.5.5 0 0 1 .183-.683zM7 6.5C7 7.328 6.552 8 6 8s-1-.672-1-1.5S5.448 5 6 5s1 .672 1 1.5zm4 0c0 .828-.448 1.5-1 1.5s-1-.672-1-1.5S9.448 5 10 5s1 .672 1 1.5z"
            />
          </svg>
        </button>
        <input
          accept="image/*"
          aria-labelledby="img-btn"
          data-testid="fileDropzone"
          id="mediaCapture"
          type="file"
        />
        <button
          aria-label="Add an image"
          class="chat-btns"
          id="img-btn"
          title="Add an image"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M21 3H3C2 3 1 4 1 5v14c0 1.1.9 2 2 2h18c1 0 2-1 2-2V5c0-1-1-2-2-2zm0 15.92c-.02.03-.06.06-.08.08H3V5.08L3.08 5h17.83c.03.02.06.06.08.08v13.84zm-10-3.41L8.5 12.5 5 17h14l-4.5-6z"
            />
          </svg>
        </button>
        <button
          class="chat-btns"
          disabled=""
          id="send-btn"
          type="submit"
        >
          SEND
        </button>
      </form>
    </div>
  </main>
</DocumentFragment>
`;
